// To generate the gRPC code, run the following command:
// python -m grpc_tools.protoc -I. --python_out=. --grpc_python_out=. comm.proto --pyi_out=.
syntax = "proto3";

service CommunicationServer {
  rpc send_data (Data) returns (Empty) {}
  rpc get_rank (Empty) returns (Rank) {}
  rpc get_model (Empty) returns (Model) {}
  rpc get_current_round (Empty) returns (Round) {}
  rpc update_port (PeerId) returns (Empty) {}
  rpc send_peer_ids (PeerIds) returns (Empty) {}
  rpc send_quorum (Quorum) returns (Empty) {}
  rpc send_finished (Rank) returns (Empty) {}
}

message Empty {}

message Model {
  bytes buffer = 1;
}

message Data {
  string id = 1;
  Model model = 2;
}

message Rank {
  int32 rank = 1;
}

message Round {
  int32 round = 1;
}

message Port {
  int32 port = 1;
}

message PeerId {
  Rank rank = 1;
  Port port = 2;
  string ip = 3;
}

message PeerIds {
  map<int32, PeerId> peer_ids = 1;
}

message Quorum {
  bool quorum = 1;
}